File: code/scenes/src/main/java/org/fbme/scenes/cells/CompositeCellAction.java
package org.fbme.scenes.cells;

import jetbrains.mps.openapi.editor.EditorContext;
import jetbrains.mps.openapi.editor.cells.CellAction;
import org.jetbrains.annotations.NotNull;

import java.util.LinkedHashSet;
import java.util.Set;
import java.util.stream.Collectors;

public class CompositeCellAction implements CellAction {

    private final Set<CellAction> myActions = new LinkedHashSet<>();

    @Override
    public String getDescriptionText() {
        if (myActions.isEmpty()) {
            return "[Invalid Composite]";
        }
        if (myActions.size() == 1) {
            return myActions.iterator().next().getDescriptionText();
        }
        return "Composite[" + myActions.stream().map(CellAction::getDescriptionText).collect(Collectors.joining(", ")) + "]";
    }

    public void add(@NotNull CellAction action) {
        myActions.add(action);
    }

    public void remove(@NotNull CellAction action) {
        myActions.remove(action);
    }

    public boolean isInvalid() {
        return myActions.isEmpty();
    }

    @Override
    public boolean executeInCommand() {
        return myActions.stream().anyMatch(it -> executeInCommand());
    }

    @Override
    public boolean canExecute(final EditorContext context) {
        return myActions.stream().anyMatch(it -> it.canExecute(context));
    }

    @Override
    public void execute(final EditorContext context) {
        myActions.forEach(it -> it.execute(context));
    }
}

File: code/scenes/src/main/kotlin/org/fbme/scenes/cells/CompositeCellAction.kt
package org.fbme.scenes.cells

import jetbrains.mps.openapi.editor.EditorContext
import jetbrains.mps.openapi.editor.cells.CellAction

class CompositeCellAction : CellAction {
    private val actions: MutableSet<CellAction> = LinkedHashSet()

    override fun getDescriptionText(): String = when (actions.size) {
        0 -> "[Invalid Composite]"
        1 -> actions.first().descriptionText
        else -> "Composite[" + actions.joinToString(", ") { it.descriptionText } + "]"
    }

    fun add(action: CellAction) {
        actions.add(action)
    }

    fun remove(action: CellAction) {
        actions.remove(action)
    }

    val isInvalid: Boolean
        get() = actions.isEmpty()

    override fun executeInCommand(): Boolean {
        return actions.any { it.executeInCommand() }
    }

    override fun canExecute(context: EditorContext): Boolean {
        return actions.any { it.canExecute(context) }
    }

    override fun execute(context: EditorContext) {
        actions.forEach { it.execute(context) }
    }
}
